firebase-firestore/src/main/java/com/google/firebase/firestore/remote/GrpcCallProvider.java

```diff
@@ -198,4 +200,71 @@ void shutdown() {
       Thread.currentThread().interrupt();
     }
   }
+
+  /**
+   * Cancels the connectivityStateTimer if the new state indicates gRPC is online. Starts the timer
+   * if the new state indicates gRPC is connecting.
+   */
+  private void onConnectivityStateChange(ManagedChannel channel) {
```

Comment:
It might be worth adding some logging to all of your changes so that we can tell from the logs when gRPC changes state and when we start / stop the timer, reset the channel, etc.  

In the short-term you can possibly use this logging to help validate these changes are behaving as expected (play around with an app and trigger different kinds of connection problems and make sure the logs make sense).  Longer-term I think it could be extremely helpful when debugging future connection-related complaints from customers.

FWIW- One thing I'm specifically curious about is if onConnectivityStateChange() will be called in response to us calling shutdownNow().  If it _does_ get called we need to make sure it doesn't cause any problems (e.g. creating an extra connection timer or something).